src/garage/torch/algos/vpg.py

```diff
@@ -191,8 +205,36 @@ def train_once(self, itr, paths):
             discount=self.discount)
         return np.mean(undiscounted_returns)
 
-    def _train(self, obs, actions, rewards, advantages):
-        r"""Train the algorithm with minibatch.
+    def _train(self, obs_flat, actions_flat, rewards_flat, returns_flat,
+               advs_flat):
+        r"""Train the policy and value function with minibatch.
+
+        Args:
+            obs_flat (torch.Tensor): Observation from the environment
+                with shape :math:`(N, O*)`.
+            actions_flat (torch.Tensor): Actions fed to the environment
+                with shape :math:`(N, A*)`.
+            rewards_flat (torch.Tensor): Acquired rewards
+                with shape :math:`(N, )`.
+            returns_flat (torch.Tensor): Acquired returns
```

Comment:
why do you need returns from the CPU? we have a function to calculate returns in the GPU.

Reply:
Can you clarify what is the meaning of "from the CPU"?
I just get values from `tensor_utils.discount_cumsum`, flatten the values, and make minibatch.
Do I need to make another function in PyTorch to compute the returns?

Reply:
the point here is that the returns can be differentiable.